### PR TITLE
[MRG+1] Addressing issue #3211

### PR DESCRIPTION
**Pull request by whalebot-helmsman**
Return non-zero exit code from commands in case if we have error in spiders constructor

### CODE REVIEW

File: scrapy/crawler.py [hunk 1]
```diff
@@ -221,6 +227,11 @@ def join(self):
         while self._active:
             yield defer.DeferredList(self._active)
 
+    def is_crawlers_has_spider(self):
+        return reduce(lambda x,y: x and y,
+                      self.crawlers_has_spiders,
+                      True)
```
Comment (kmike) [inline scrapy/crawler.py]: `any(self.crawlers_has_spiders)`?

Comment (kmike) [reply scrapy/crawler.py]: as @VMRuiz noted, it is actually `all(...)`, not `any(...)`, which is an additional argument to change it, as the code is not obvious :P

**Comment by whalebot-helmsman** [comment]
Test failed due to new twisted release, fix in twisted trunk https://github.com/twisted/twisted/pull/1008 but not released to pypy

File: tests/test_commands.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+import pdb
+import inspect
```
Comment (kmike) [inline tests/test_commands.py]: these imports are unneeded

File: tests/test_crawler.py [hunk 3]
```diff
@@ -11,7 +11,10 @@
 from scrapy.utils.log import configure_logging, get_scrapy_root_handler
 from scrapy.utils.spider import DefaultSpider
 from scrapy.utils.misc import load_object
+from scrapy.utils.test import get_crawler
 from scrapy.extensions.throttle import AutoThrottle
+from twisted.internet import defer
+import twisted.trial.unittest
```
Comment (kmike) [inline tests/test_crawler.py]: as per pep8, twisted imports should go in a separate group, between stdlib imports and scrapy imports

File: tests/test_commands.py [hunk 4]
```diff
@@ -220,6 +222,16 @@ def test_runspider(self):
         self.assertIn("INFO: Closing spider (finished)", log)
         self.assertIn("INFO: Spider closed (finished)", log)
 
+    def test_run_fail_spider(self):
+        proc = self.runspider("import scrapy\n" + inspect.getsource(ExceptionSpider))
```
Comment (kmike) [inline tests/test_commands.py]: nice trick!

**Comment by kmike** [comment]
Looks good to me, thanks @whalebot-helmsman! I'd remove `_is` from method names, but it is just a style preference. +1 to merge.

File: scrapy/crawler.py [hunk 5]
```diff
@@ -221,6 +226,9 @@ def join(self):
         while self._active:
             yield defer.DeferredList(self._active)
 
+    def _is_spider_created_for_every_crawler(self):
+        return all(self.crawlers_has_spiders)
```
Comment (dangra) [inline scrapy/crawler.py]: I'd like to propose the removal of all this methods and use a single boolean flag that summarizes the outcome of bootstraping the spider. 

The public attribute name would be `bootstrap_failed` and we set it at `_crawl / _done()` like:
```python
    self.bootstrap_failed |= not getattr(crawler, 'spider', None)
```
this public attribute replaces `_is_spider_created_for_every_crawler()` method and `crawlers_has_spiders` atttribute, but also remove the need for private `_is_spider_created()` method.   

then the check in `runspider.py` and `crawl.py` looks like:

```python
if not self.crawler_process.bootstrap_failed:
    self.exitcode = 1
```

**Comment by kmike** [comment]
Thanks @whalebot-helmsman and @dangra!

**Comment by dangra** [comment]
thanks @whalebot-helmsman and @kmike

**Comment by dangra** [comment]
I restarted the failed build to be sure it was a temporal failure.

**Pull request merged**